### PR TITLE
Fixing 0 length issue with constructor

### LINKED ISSUE #106538: System.Numerics.Tensors Non-monotonic constructor errors
### Description

```csharp
new TensorSpan<double>(new double[j], lengths: new IntPtr[] {2}, strides: new IntPtr[] {1}); 
```
throws a ArgumentException if `j == 1` but not if `j == 0` or `j > 1`.

### Reproduction Steps

```csharp
using System.Diagnostics.CodeAnalysis;
using System.Numerics.Tensors;

namespace ConsoleApp1;

public abstract class Program
{
	[Experimental("SYSLIB5001")]
	public static void Main()
	{
		new TensorSpan<double>(new double[0], lengths: new IntPtr[] {2}, strides: new IntPtr[] {1});
		new TensorSpan<double>(new double[2], lengths: new IntPtr[] {2}, strides: new IntPtr[] {1});
		new TensorSpan<double>(new double[1], lengths: new IntPtr[] {2}, strides: new IntPtr[] {1}); 
	}
}
```

### Expected behavior

Error when `j <= 1`

### Actual behavior

Error if and only if `j == 1`

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

.NET8, Ubuntu22, x64

### Other information

_No response_

### PR DESCRIPTION
**Pull request by michaelgsharp**
Fixes #106538 by modifying the check in the constructor to also throw an error if the base memory length is 0.

### CODE REVIEW

**Comment by rameel** [comment]
This expression can be simplified.

```csharp
if ((maxElements >= span.Length && span.Length != 0) || (span.Length == 0 && maxElements > 0))
    ThrowHelper.ThrowArgument_InvalidStridesAndLengths();
```

For simplicity, let's represent `maxElements` as `a` and `span.Length` as `b`. Now, the condition looks like this:
```csharp
if ((a >= b && b != 0) || (b == 0 && a > 0))
    error;
```

We can normalize the condition by switching the order of the expressions `(b == 0 && a > 0)`, resulting in:
```csharp
if ((a >= b && b != 0) || (a > 0 && b == 0))
    error;
```

This can be interpreted as:
```csharp
if ((a >= b && b != 0)  // a > 0, b > 0, a >= b
 || (a > 0  && b == 0)) // a > 0, b = 0
    error;
```

Thus, we are only interested in cases where `a > 0`, which simplifies the second part of the expression.

In the end, we get:
```csharp
if (a > 0 && a >= b)
    error;
```

The final expression will look like this:
```csharp
if (maxElements > 0 && maxElements >= span.Length)
    ThrowHelper.ThrowArgument_InvalidStridesAndLengths();
```

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/ReadOnlyTensorSpan.cs [hunk 1]
```diff
@@ -129,7 +129,7 @@ public ReadOnlyTensorSpan(ReadOnlySpan<T> span, scoped ReadOnlySpan<nint> length
             strides = strides.IsEmpty ? (ReadOnlySpan<nint>)TensorSpanHelpers.CalculateStrides(lengths, linearLength) : strides;
             TensorSpanHelpers.ValidateStrides(strides, lengths);
             nint maxElements = TensorSpanHelpers.ComputeMaxLinearIndex(strides, lengths);
-            if (maxElements >= span.Length && span.Length != 0)
+            if ((maxElements >= span.Length && span.Length != 0) || (span.Length == 0 && maxElements > 0))
```
Comment (stephentoub) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/ReadOnlyTensorSpan.cs:132]:
```suggestion
            if (span.IsEmpty ? maxElements > 0 : maxElements >= span.Length)
```
?

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/ReadOnlyTensorSpan.cs [hunk 2]
```diff
@@ -129,7 +129,7 @@ public ReadOnlyTensorSpan(ReadOnlySpan<T> span, scoped ReadOnlySpan<nint> length
             strides = strides.IsEmpty ? (ReadOnlySpan<nint>)TensorSpanHelpers.CalculateStrides(lengths, linearLength) : strides;
             TensorSpanHelpers.ValidateStrides(strides, lengths);
             nint maxElements = TensorSpanHelpers.ComputeMaxLinearIndex(strides, lengths);
-            if (maxElements >= span.Length && span.Length != 0)
+            if (span.IsEmpty ? maxElements > 0 : maxElements >= span.Length)
```
Comment (tannergooding) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/ReadOnlyTensorSpan.cs:132]: Does `ComputeMaxLinearIndex` throw on overflow? That is, is `maxElements` guaranteed to be non-negative, in which case `maxElements != 0` might be clearer?

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorSpan.cs [hunk 3]
```diff
@@ -132,7 +132,7 @@ public TensorSpan(Span<T> span, scoped ReadOnlySpan<nint> lengths, scoped ReadOn
             TensorSpanHelpers.ValidateStrides(strides, lengths);
 
             nint maxElements = TensorSpanHelpers.ComputeMaxLinearIndex(strides, lengths);
-            if (maxElements >= span.Length && span.Length != 0)
+            if ((maxElements >= span.Length && span.Length != 0) || (span.Length == 0 && maxElements > 0))
```
Comment (tannergooding) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorSpan.cs:135]: Why the different checks in `TensorSpan` vs `ReadOnlyTensorSpan`?

Comment (michaelgsharp) [reply src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorSpan.cs:135]: Fixed.

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/ReadOnlyTensorSpan.cs (hunk 2 above)
Comment (michaelgsharp) [reply src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/ReadOnlyTensorSpan.cs:132]: `ComputeMaxLinearIndex` does not throw, but we validate and throw if a length is negative before this point, so `maxElements` is guaranteed to be non-negative.

Comment (michaelgsharp) [reply src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/ReadOnlyTensorSpan.cs:132]: Fixed.

**Review by tannergooding — APPROVED** [review]

**Pull request merged**